modules/core/src/main/java/org/apache/ignite/internal/processors/datastructures/GridCacheSemaphoreImpl.java

```diff
@@ -662,6 +665,28 @@ private void initializeSemaphore() throws IgniteCheckedException {
         }
     }
 
+    /** {@inheritDoc} */
+    @Override public <T> IgniteFuture<T> acquireAndExecute(IgniteCallable<IgniteFuture<T>> callable,
+                                                           int numPermits) throws Exception {
+        acquire(numPermits);
+
+        IgniteFuture<T> future = null;
+        try {
+            future = callable.call();
+        } finally {
+            release(numPermits);
```

Comment:
Is this release required? Since `call()` here is asynchronous, permits will be released before completion, which seems wrong. Also, there is another `release` invocation, which happens within the future listener - that one looks correct.